proposals/p1928.md

```diff
@@ -0,0 +1,163 @@
+
+## Proposal
+
+This proposal introduces array internal representation, its declaration, and
+indexing.
+
+## Details
+
+### Internal representation
+
+For simplifying usage and raising safety of array usage, internal representation
+of arrays contains additional metadata:
+```cpp
+  struct {
+    int32_t dimensions;
+    int32_t * dimension_size;
+    void * data;
+  } Array;
+```
+Here, `dimensions` contains information about number of dimensions (necessary
+for multidimensional arrays), `dimension_size` is an array of size `dimensions`
+containing sizes for each dimensions, and `data` is continuous in memory place
+keeping scalar data.
+
+This representation is closer to `std::vector<T>` rather then `T[]`. Then,
+interoperability between C++ and carbon is limited.
+
+One-dimensional and nested arrays have `dimensions = 1`.
+
+### Declaration
+
+During declaration of arrays, it is more importantly to see type of variable.
+It is the aim of the below presented alternatives of declaration.
+
+Current realization:
+```carbon
+  var array_1D: [i32; 4];
+  var array_2D: [i32; 4, 10]; // is not supported yet
+  var array_nested: [[i32; 10]; 4];
+```
+In the case of nested array, definition of them a bit misunderstood due to
+following indexing is inversed.
+
+Suggested in [#1787](https://github.com/carbon-language/carbon-lang/pull/1787):
+```carbon
+  var array_1D: [4; i32];
+  var array_2D: [4, 10; i32];
+  var array_nested: [4; [10; i32]];
+```
+This definition is deprived of a defect of previous definition, but the type is
+the last one.
+
+Alternatives:
+```carbon
+  var array_1D[4]: i32;
+  var array_2D[4, 10]: i32;
+  var array_nested[4][10]: i32;
+```
+In this alternative, `auto` will work only for types, not for dimensions of
+array.
+
+```carbon
+  var array_1D: i32[4];
+  var array_2D: i32[4, 10];
+  var array_nested: i32[4][10];
+```
+This definition requires contextual information since `i32[4]` is the same as
+`array_1D[1]` so it is necessary to separate type and variable.
+
+Also, it is possible to declare array in function with runtime defined size:
+```carbon
+  fn f(n: i32) -> i32 {
+    var arr: [n; i32];
+    return 0;
+  }
+```
+
+Only of size is defined in compile-time and it is less than
+`MAX_STACK_SIZE_ARRAY`, array is located in stack memory. Other definitions are
+located in heap memory. It is especially important for the last example.
+
+### Indexing
+
+Indexing possible via integer values like in C style:
+```carbon
+  var array: [4; i32];
+  var elem: auto = array[1];
+```
+and also via array:
+```carbon
+  var array_ind: [3; i32] = (0, 2, 4);
+  var array: [5; i32] = (4, 3, 2, 1, 0);
+  var arrav: auto = array[array_ind];
+  // arrav == (array[0], array[2], array[4]) == (4, 2, 0)
+```
+Array indexing is useful if slices generate arrays.
+
+Indexes are also possible on left-side expression:
+```carbon
+  var array: [6; i32] = (5, 4, 3, 2, 1, 0);
+  array[(0,1,2)] = array[(3,4,5)]; // array = (2, 1, 0, 2, 1, 0)
```

Comment:
What are the types of the left and right and side?

Reply:
left side is `[3; i32]` and right --- `[3; i32]`.

Reply:
I haven't seen that syntax before, does that syntax exist in a different language?
It feels a bit unclear to me, and seems to force users to manually indicate individual indexes. Supporting ranges of indexes (for example like Python does `slice[begin:end]`) would be super valuable for sure.

Reply:
Can I write `array[(0,2,4)] = array[(1,3,5)]`?
If so, I don't think the left-hand side can be an array. It would need to store references to each particular element that can be assigned to.

But to be honest, I'm not sure this passes the readability test for me (maybe it's just lack of familiarity?) It doesn't seem too bad to me to write
`(array[0], array[2], array[4]) = (array[1], array[3], array[5])`, or just three assignments.

Reply:
> I haven't seen that syntax before, does that syntax exist in a different language?

It exists in Fortran; see section 6.2.2.3.2 Vector subscript in [Draft of Fortran 2003 Standard](https://j3-fortran.org/doc/year/04/04-007.pdf).

> Supporting ranges of indexes <...> would be super valuable for sure.

See #1927. Slices will generate arrays then they may be used as indexes, so `array[begin:end]` will be possible with two these proposals.

> Can I write array[(0,2,4)] = array[(1,3,5)]?

Yes, you can. I checked what Fortran compiler do: it is just a simplification of cycles. The main idea why array should be indexes via arrays is slice usage.

Reply:
I agree with @asoffer regarding the readability of `array[(0,2,4)] = array[(1,3,5)]`. To me it looks like we build arrays from tuples, or similar. I don't think we should allow (or at least promote) this kind of syntax. `(array[0], array[2], array[4]) = (array[1], array[3], array[5])` looks fine to me as well.

Reply:
What would `array[(0,1)] = array[(1,0)];` do? Is that a swap, or does one of the elementwise assignments happen first? If it's a swap, then in general (with runtime indexes on both sides) would we form an intermediate array with the selected elements?